pkg/operator/controller/externaldns/deployment.go

```diff
@@ -448,6 +477,28 @@ func externalDNSVolumesChanged(current, expected, updated *appsv1.Deployment) bo
 	return changed
 }
 
+// externalDNSDeploymentStrategyChanged returns true if the current update strategy differs from expected.
+func externalDNSDeploymentStrategyChanged(current, expected, updated *appsv1.Deployment) bool {
+	changed := false
+
+	// reset the strategy type if it differs
+	if expected.Spec.Strategy.Type != "" &&
+		expected.Spec.Strategy.Type != current.Spec.Strategy.Type {
+		updated.Spec.Strategy.Type = expected.Spec.Strategy.Type
+		changed = true
+	}
```

Comment:
I'm not resetting the whole strategy for the same reason as I'm not setting the replicas field if it's `nil`: to not fight with the api server defaulting.  For the recreate strategy it's less important as it doesn't have any fields other than `Type` however resetting the whole strategy may give a wrong impression that all the types are handled. However for the rolling strategy where `MaxUnavailable` and `MaxSurge` have the defaults we risk to get into the fighting loop if any of them is left unset in the desired state.

Reply:
You should make the changes to the API to enforce the recreate strategy if you think there's going to be flapping to the defaults.

Reply:
About the flapping: I was referring to the potential future improvements (Rolling strategy as desired). For the recreate strategy the resetting of the type is enough as it's the only field in it. BTW which API you were thinking I could make changes to?

Reply:
I suppose the same one you referred to when you said "...to not fight with the api server defaulting."  I still don't understand why you need to fight with the api server defaults, tbh.  I still think you should do it the suggested way to make sure there are no dangling parameters that you have to take care of in the next step.

Reply:
I don't think this is the same.  I am just asking you to reset the updated.Spec.Strategy, which includes the Type, instead of just the Type.  The only other parameter is called RollingUpdate, in https://github.com/openshift/external-dns-operator/blob/0cd8bee31e64f3086f9a44e3f5e299a3a1111175/vendor/k8s.io/api/apps/v1/types.go#L407 .  I am asking you to remove the RollingUpdate parameter instead of potentially leaving it behind when you reset the Type from RollingUpdate to Recreate.  It doesn't make sense to leave the RollingUpdate parameter around if you reset the Type from RollingUpdate to Recreate.

Reply:
Yes, I agree the rolling update parameters have to be cleaned and this is what I already do [here](https://github.com/openshift/external-dns-operator/blob/d8e0462d71ecb2d5fc30e45cc9d86e1c082f8d0d/pkg/operator/controller/externaldns/deployment.go#L491-L497). There is a [dedicated unit test case](https://github.com/openshift/external-dns-operator/blob/d8e0462d71ecb2d5fc30e45cc9d86e1c082f8d0d/pkg/operator/controller/externaldns/deployment_test.go#L6008-L6028) for it too.